### ISSUE TITLE
[Errno 32] Broken pipe for Parser in parallel execution on OSX

### ISSUE THREAD

**Issue by mladvladimir** [HazyResearch/fonduer#47]
Hi,

In [fonduer-tutorials](https://github.com/HazyResearch/fonduer-tutorials), after running cell:
```python
corpus_parser = OmniParser(structural=True, lingual=True, visual=True, pdf_path=pdf_path)
%time corpus_parser.apply(doc_preprocessor, parallelism=PARALLEL)
```

whenever is PARALLEL smaller than max_docs, I've got:
```
Traceback (most recent call last):
  File "/anaconda3/lib/python3.6/multiprocessing/queues.py", line 240, in _feed
    send_bytes(obj)
  File "/anaconda3/lib/python3.6/multiprocessing/connection.py", line 200, in send_bytes
    self._send_bytes(m[offset:offset + size])
  File "/anaconda3/lib/python3.6/multiprocessing/connection.py", line 398, in _send_bytes
    self._send(buf)
  File "/anaconda3/lib/python3.6/multiprocessing/connection.py", line 368, in _send
    n = write(self._handle, buf)
BrokenPipeError: [Errno 32] Broken pipe
```

Otherwise (with PARALLEL bigger or equal than max_docs) result is empty tables in Postgresql. 
When turning off parallelisation, it works.

Best regards

**Comment by SenWu** [comment]
Hi @mladvladimir, Thanks for reporting this issue. Unfortunately, we did see this issue on our side. Can you give us more information about your environment and setup?

**Comment by lukehsiao** [comment]
Hi @mladvladimir 

Can you try a fresh install using brew for postgres and using a virtualenv rather than conda?

**Comment by lukehsiao** [comment]
@mladvladimir Also, did you make sure to both create the db used in the tutorial, and download the tutorial data before running the tutorial?

**Comment by mladvladimir** [comment]
@lukehsiao 

I tried again. 
New version of Postgresql is installed with brew (PostgreSQL 10.3);
virtualenv is created without anaconda; 
database is created and data files downloaded according to tutorial; 
storage schema  was initialized successfully (tables are created). 
HTMLPreprocessor is defined without problems.
It failing constantly on  `corpus_parser.apply(doc_preprocessor, parallelism=PARALLEL)` step.
What is even more strange that all other steps which use UDFRunner with parallelism (CandidateExtractor, BatchFeatureAnnotator) are executed without problems. 
I tried partially **max_storage_temp_tutorial** on ubuntu 18.04 Docker and it's succeed in mode (structural=True, lingual=True, visual=False) but  error remains if all tree parameters are True.
Is there any particular version of python3 that you recommend?
 
By the way, in shouldn't be `parralelism` instead of `parallel` in test_parser.py?https://github.com/HazyResearch/fonduer/blob/84364d8076153775e02f0a0e12d060c44eae3733/tests/parser/test_parser.py#L189 
This makes that test run in single threaded mode.

Generally speaking, complete project is precious :)

Regards
Vladimir

**Comment by lukehsiao** [comment]
Unfortunately, I'm still not sure quite sure what is wrong. I haven't personally tested in ubuntu 18.04 either, we have mostly been using 16.04, as you can see in our e2e Travis test and in the tutorials. If you have any more error logs or information that might help us recreate the issue, we can keep looking at it.

> Is there any particular version of python3 that you recommend?

We have tested with Python 3.5 and 3.6.

> By the way, in shouldn't be `parralelism` instead of `parallel` in test_parser.py?

 Nice catch! I've opened a new PR (#48) to fix that.

**Comment by lukehsiao** [comment]
We've been able to reproduce this and are looking into it.

**Comment by mladvladimir** [comment]
Thanks for info;
After some observing, it seems like Spacy pipeline call is problematic with parallelism:
https://github.com/HazyResearch/fonduer/blob/066b01c296c6f02b0b312f1170498b02da4cc780/fonduer/parser/spacy_parser.py#L128-L129

maybe I'm wrong.

**Comment by lukehsiao** [comment]
This may be related: https://github.com/explosion/spaCy/issues/1572

**Comment by sivasrc** [comment]
Hi @SenWu,

When I try execute the given sample 
%time corpus_parser.apply(doc_preprocessor,parallelism=PARALLEL), getting the below issue, not sure why, any help would be appreciated.

[INFO] fonduer.utils.udf - Running UDF...
0% 0/78 [00:00<?, ?it/s]
---------------------------------------------------------------------------
SystemExit                                Traceback (most recent call last)
<timed eval> in <module>

~/.local/lib/python3.6/site-packages/fonduer/parser/parser.py in apply(self, doc_loader, pdf_path, clear, parallelism, progress_bar)
    108             clear=clear,
    109             parallelism=parallelism,
--> 110             progress_bar=progress_bar,
    111         )
    112 

~/.local/lib/python3.6/site-packages/fonduer/utils/udf.py in apply(self, doc_loader, clear, parallelism, progress_bar, **kwargs)
     70             self._apply_st(doc_loader, clear=clear, **kwargs)
     71         else:
---> 72             self._apply_mt(doc_loader, parallelism, clear=clear, **kwargs)
     73 
     74         # Close progress bar

~/.local/lib/python3.6/site-packages/fonduer/utils/udf.py in _apply_mt(self, doc_loader, parallelism, **kwargs)
    128                 out_queue=out_queue,
    129                 worker_id=i,
--> 130                 **self.udf_init_kwargs
    131             )
    132             udf.apply_kwargs = kwargs

~/.local/lib/python3.6/site-packages/fonduer/parser/parser.py in __init__(self, structural, blacklist, flatten, lingual, strip, replacements, tabular, visual, pdf_path, language, **kwargs)
    169         if self.lingual_parser.has_tokenizer_support():
    170             self.tokenize_and_split_sentences = self.lingual_parser.split_sentences
--> 171             self.lingual_parser.load_lang_model()
    172         else:
    173             self.tokenize_and_split_sentences = SimpleTokenizer().parse

~/.local/lib/python3.6/site-packages/fonduer/parser/spacy_parser.py in load_lang_model(self)
    125         if self.lang in self.languages:
    126             if not Spacy.model_installed(self.lang):
--> 127                 download(self.lang)
    128             model = spacy.load(self.lang)
    129         elif self.lang in self.alpha_languages:

~/.local/lib/python3.6/site-packages/spacy/cli/download.py in download(model, direct, *pip_args)
     36                             .format(m=model_name, v=version), pip_args)
     37         if dl != 0:  # if download subprocess doesn't return 0, exit
---> 38             sys.exit(dl)
     39         try:
     40             # Get package path here because link uses

SystemExit: 1

**Comment by SenWu** [comment]
Hi @sivasrc, it seems like your issue is related to downloading spacy language module. I suggest you try it again or download the module manually (here is the reference: https://spacy.io/usage/models).

**Comment by SenWu** [comment]
This issue is fixed by Spacy and test it in https://github.com/HazyResearch/fonduer/pull/176.

**Comment by sivasrc** [comment]
Thanks for your response.

I would like to locate the section, sub-section (level 2,3,4 etc) and
sentences beneath, and relate them using Fonduer.. can I achieve it?
